[HUDI-1486] Remove inflight rollback in hoodie writer

n3nash opened the pull request:
1. Refactor rollback and move cleaning failed commits logic into cleaner
2. Introduce hoodie heartbeat to ascertain failed commits
3. Fix test cases

## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+    this.fs = fs;
+    this.basePath = basePath;
+    this.heartBeatFolderPath = HoodieTableMetaClient.getHeartbeatFolderPath(basePath);
+    this.heartbeatIntervalInMillis = heartbeatIntervalInSeconds * 1000L;
```
why not just keep the config at ms level. I think thats better to do .

n3nash replied:
I think it's easier for people to reason with seconds and not set too low a value like 1000ms etc which can overwhelm RPC's

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+        LOG.warn("Stopping heartbeat for previous instant time " + this.currentInstantTime);
+        this.stop(this.currentInstantTime);
+      }
+      this.currentInstantTime = instantTime;
```
so this is assuming the single write client won't have multiple writers. i.e will issue `start(instantTime)` for only one instant at any time. Can we use a Map per instantTime and maintain state such that the heartbeatClient can handle multiple instant times at once

n3nash replied:
Refactored.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to generate heartbeat");
+    }
+    this.heartBeatFuture = executorService.submit(() -> {
```
can we use a TimerTask or TimerService. it can also support multiple of these schedules

n3nash replied:
Added.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+  private Long lastHeartBeatTime;
+  // This is required for testing. Problem : if we set heartbeatIntervalInSeconds really large, test takes longer, if
+  // we set it small and you are debugging with breakpoint, causes issues. Need to implement a Mock in tests
+  private final Boolean skipHeartBeatCheck;
```
this is always true?

n3nash replied:
Remnant of testing, remove it

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+        }
+      }
+    });
+    this.isHearbeatStarted = true;
```
then all we need to is to cancel the timers

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+    }
+  }
+
+  public boolean checkIfConcurrentWriterRunning(String instantTime) throws IOException {
```
this is leaking upper level context (multi writing) to a lower level class. We should just name this something around heartbeats. say `isHeartbeatExpired()`

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+  public boolean checkIfConcurrentWriterRunning(String instantTime) throws IOException {
+    long lastHeartBeatForWriter = getLastHeartbeat(fs, basePath, instantTime);
+    long currentTime = System.currentTimeMillis();
+    if (currentTime - lastHeartBeatForWriter > this.heartbeatIntervalInMillis) {
```
`return this.heartbeatIntervalInMillis <= currentTime - lastHeartBeatForWriter` ? single line

vinothchandar replied:
So the heartbeats won't align correctly with the interval. Typically we add another config that multiplies with `this.heartbeatIntervalInMillis` , which should at least be 2. I am saying we should compared for `currentTime - lastHeartBeatForWriter > this.numTolerableHeartbeatMisses * this.heartbeatIntervalInMillis`.

`numTolerableHeartbeatMisses` should have another config, with default value 2

n3nash replied:
I want to add logs if the heartbeat has expired, hence not in a single line. Added the other property for tolereable misses

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,196 @@
+
+  private void updateHeartbeat(String instantTime) throws IOException {
+    Long newHeartBeatTime = System.currentTimeMillis();
+    OutputStream outputStream =
```
AFAIK, a new file creation should be atomic in both HDFS and cloud stores. but unsure how the overwrite part works. 

Can you clarify?

n3nash replied:
Yes, a new file creation is atomic, overwrite is NOT. That is why I don't read the contents of the file since that causes race-conditions and the attempt to read fail, the modification time is an indicator of when was the last attempt to update heartbeat which is all we need. Also from the documentation here -> https://hadoop.apache.org/docs/r3.1.2/hadoop-project-dist/hadoop-common/filesystem/introduction.html
`When OutputStream.close() is called, all remaining data is written, the file closed and the NameNode updated with the final size of the file. The modification time is set to the time the file was closed.`

vinothchandar commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/heartbeat/TestHoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,160 @@
+        new HoodieHeartbeatClient(metaClient.getFs(), metaClient.getBasePath(), heartBeatInterval);
+    hoodieHeartbeatClient.start("100");
+    // wait 10 heartbeats
+    Thread.sleep(10 * heartBeatInterval * 1000);
```
No `Thread.sleep` in tests please. these make tests flaky. lets structure them, to wait till a condition X is reached with a timeout. essence is explained here. http://www.awaitility.org/

vinothchandar commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/heartbeat/TestHoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,160 @@
+          System.out.println("Modification Time => " + fs.getModificationTime());
+        }
+      } catch (Exception e) {
+        System.out.println("Caught Exception th2");
```
log?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -100,7 +107,9 @@ public static SparkConf registerClasses(SparkConf conf) {
     List<HoodieWriteStat> writeStats = writeStatuses.map(WriteStatus::getStat).collect();
-    return commitStats(instantTime, writeStats, extraMetadata, commitActionType, partitionToReplacedFileIds);
+    boolean isCommitted = commitStats(instantTime, writeStats, extraMetadata, commitActionType, partitionToReplacedFileIds);
+    this.hoodieHeartbeatClient.stop(instantTime);
```
hmmm. we should probably do this only once at the super class level.

n3nash replied:
Moved to `postCommit`

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -437,7 +456,10 @@ private void testUpsertsInternal(HoodieWriteConfig config,
   @Test
   public void testDeletes() throws Exception {
-    SparkRDDWriteClient client = getHoodieWriteClient(getConfig(), false);
+    // Set cleaner to LAZY so no inflights are cleaned
```
why can't we override this once for all tests. if we run everything with `LAZY` then we can have better confidence that no uncommitted data will ever be exposed

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -546,7 +571,10 @@ private void testUpsertsUpdatePartitionPath(IndexType indexType, HoodieWriteConf
         metaClient.getTableConfig().getPayloadClass(), VERSION_0);
-    SparkRDDWriteClient client = getHoodieWriteClient(hoodieWriteConfig, false);
+    // Set rollback to LAZY so no inflights are deleted
+    hoodieWriteConfig.getProps().put(HoodieCompactionConfig.FAILED_WRITES_CLEANER_POLICY_PROP,
```
We need to do something about the repeated line

`getConfig().getProps().put(HoodieCompactionConfig.FAILED_WRITES_CLEANER_POLICY_PROP,
        HoodieCleaningPolicy.HoodieFailedWritesCleaningPolicy.LAZY.name());`

even if you want to override only selectively. Can we add an override for `getConfig(boolean)` which decides whether or not we do lazy cleaning of failed writes.

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -1367,6 +1410,77 @@ public void testRollbackAfterConsistencyCheckFailureUsingMarkers(boolean enableO
+  @Test
+  public void testParallelWriting() throws Exception {
+    ExecutorService service = Executors.newFixedThreadPool(3);
+    String instantTime1 = "100";
```
lets inline these variables . they can all be just constants, right

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -1367,6 +1410,77 @@ public void testRollbackAfterConsistencyCheckFailureUsingMarkers(boolean enableO
+
+    List<HoodieRecord> records0 = dataGen.generateInserts(instantTime, 200);
+    JavaRDD<HoodieRecord> writeRecords0 = jsc.parallelize(records0, 1);
+    JavaRDD<WriteStatus> result0 = client.bulkInsert(writeRecords0, instantTime);
```
should we upsert of the writes. so we get a better mix.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieCleaningPolicy.java:
```diff
@@ -22,5 +22,8 @@
 public enum HoodieCleaningPolicy {
-  KEEP_LATEST_FILE_VERSIONS, KEEP_LATEST_COMMITS
+  KEEP_LATEST_FILE_VERSIONS, KEEP_LATEST_COMMITS;
+  public enum HoodieFailedWritesCleaningPolicy {
```
let's have as a top level enum/ unnested

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/MultiWriterConcurrencyModel.java:
```diff
@@ -0,0 +1,25 @@
+
+package org.apache.hudi.common.model;
+
+public enum MultiWriterConcurrencyModel {
```
rename:WriteConcurrencyMode

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/MultiWriterConcurrencyModel.java:
```diff
@@ -0,0 +1,25 @@
+package org.apache.hudi.common.model;
+
+public enum MultiWriterConcurrencyModel {
+  LEGACY_SINGLE_WRITER,
```
rename: SINGLE_WRITER, dropping the LEGACY. The model has its merits

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/MultiWriterConcurrencyModel.java:
```diff
@@ -0,0 +1,25 @@
+
+public enum MultiWriterConcurrencyModel {
+  LEGACY_SINGLE_WRITER,
+  EMBARRASSINGLY_PARALLEL,
```
rename: just `PARALLEL_NO_CONTENTION`

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -217,6 +218,13 @@ public String getMetaAuxiliaryPath() {
+  /**
+   * @return Heartbeat folder path.
+   */
+  public static String getHeartbeatFolderPath(String basePath) {
```
should nt this be under the metaPath?

n3nash replied:
It should be and is, follows the same semantic of .aux -> https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java#L76

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordScanner.java:
```diff
@@ -145,6 +146,11 @@ public void scan() {
         }
+        if (r.getBlockType() != CORRUPT_BLOCK && r.getBlockType() != COMMAND_BLOCK
+            && !timeline.containsOrBeforeTimelineStarts(r.getLogBlockHeader().get(INSTANT_TIME))) {
+          // hit an invalid block possibly from a failed write, move to the next one and skip processing this one
```
this is not an invalid block right. this skipping an uncommitted block?

n3nash replied:
Yes, that's why in the comments added "possibly from a failed write". Changed the invalid -> uncommitted

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieDefaultTimeline.java:
```diff
@@ -107,7 +107,7 @@ public HoodieTimeline filterCompletedAndCompactionInstants() {
 
   @Override
-  public HoodieDefaultTimeline getCommitsAndCompactionTimeline() {
+  public HoodieDefaultTimeline getCommitsCompactionAndReplaceTimeline() {
```
got it. so this is a rename. discard my earlier comment.

vinothchandar replied:
how about `getWriteTimeline` simply

n3nash replied:
sg

vinothchandar commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/FileCreateUtils.java:
```diff
@@ -83,6 +83,17 @@ private static void createMetaFile(String basePath, String instantTime, String s
     }
   }
 
+  private static void createMetaFile(FileSystem fs, String basePath, String instantTime, String suffix) throws IOException {
```
dont we already have helpers that do these. cc @xushiyan

xushiyan replied:
`HoodieTestTable` makes use of this `FileCreateUtils` to perform the actual file manipulations.

xushiyan replied:
Looks like this new set of helpers is a version that uses hadoop fs instead of java nio. Would it align better if we put `FileSystem fs` as the last argument? so that other args match with the java nio version. This applies to the few new helper methods below.

vinothchandar replied:
+1

vinothchandar commented on hudi-hadoop-mr/src/test/java/org/apache/hudi/hadoop/TestHoodieROTablePathFilter.java:
```diff
@@ -84,7 +84,6 @@ public void testNonHoodiePaths() throws IOException {
   }
 
-  @Test
   public void testPartitionPathsAsNonHoodiePaths() throws Exception {
```
why removing the test?

n3nash replied:
Added it back, was failing so had commented it out

vinothchandar reviewed (commented):
Will review the cleaning/archival changes more closely again later today

vinothchandar commented:
also, the writer should check if heartbeat has expired before commiting.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -165,13 +167,17 @@ public boolean archiveIfRequired(HoodieEngineContext context) throws IOException
         table.getActiveTimeline().filterPendingCompactionTimeline().firstInstant();
+    Option<HoodieInstant> oldestInflightInstant =
+        table.getActiveTimeline()
+            .getTimelineOfActions(CollectionUtils.createSet(HoodieTimeline.COMMIT_ACTION, HoodieTimeline.DELTA_COMMIT_ACTION))
```
is n't there an existing method to get this? Also should we include `replacecommit` here as well?  I understand the issue with compactions, since the retry will attempt to reuse the same instantTime. is that why you only have commit/deltacommit here?

n3nash replied:
Yes, the retry of both will use the same instant right now. All of the lazy cleaning is only done for commits/deltacommits since there is contention when multiple writers start running. Both compaction and clustering will continue to do inline rollbacks since both these operations are bound to happen once scheduled. If the compaction or clustering DON'T run, the failed writes will keep lying around. Basically, once a compaction/clustering is scheduled, we ensure no other writer can modify the same set of files, there is no contention for the failed writes and hence the cleanup happens inline. If they don't happen inline and depend on the LAZY cleaning, then since we re-use the same instant, we cannot guarantee correctness.

vinothchandar replied:
makes sense to have the lazy cleaning for just the true write actiosn.

vinothchandar reviewed (commented):
In general, I have a concern around compactions. it's possible for a compaction (i.e .commit action) to be attempted and then reattempted after a long time. In this scenario, the heartbeats will expire and we will clean it as a failed compaction?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -180,7 +186,14 @@ public boolean archiveIfRequired(HoodieEngineContext context) throws IOException
-          }).limit(commitTimeline.countInstants() - minInstantsToKeep);
+          });
+      if (config.getFailedWritesCleanPolicy() == HoodieCleaningPolicy.HoodieFailedWritesCleaningPolicy.LAZY
+          && config.getMultiWriterConcurrencyModel() != MultiWriterConcurrencyModel.LEGACY_SINGLE_WRITER) {
```
do we need to guard this by `config.getMultiWriterConcurrencyModel() != MultiWriterConcurrencyModel.LEGACY_SINGLE_WRITER` ? it should technically work right? if so, best to keep just the one flag to fence

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/BaseRollbackActionExecutor.java:
```diff
@@ -130,23 +131,29 @@ private void validateSavepointRollbacks() {
+    // since with LAZY rollback we support parallel writing which can allow a new inflight while rollback is ongoing
+    // Remove this once we support LAZY rollback of failed writes by default as parallel writing becomes the default
+    // writer mode.
+    if (config.getFailedWritesCleanPolicy() == HoodieCleaningPolicy.HoodieFailedWritesCleaningPolicy.EAGER) {
```
just return if the policy is not EAGER?  easier to read that, than nested if blocks .

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/clean/SparkCleanActionExecutor.java:
```diff
@@ -23,9 +23,9 @@
+import org.apache.hudi.common.HoodieCleanStat;
 import org.apache.hudi.common.model.CleanFileInfo;
 import org.apache.hudi.common.model.HoodieKey;
 import org.apache.hudi.common.model.HoodieRecord;
```
I was expecting some changes in this file, to clean the failed writers for LAZY? Can you point me to the files that do this.

n3nash replied:
@vinothchandar No change in the action executor, to keep this change simple, just added that here -> https://github.com/apache/hudi/blob/35a82c7f68c8bf0b381e47dc3f12374bec7081eb/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java#L609

vinothchandar commented:
@n3nash I have left my first round of comments. 

Main call out: https://github.com/apache/hudi/pull/2359#discussion_r549550003

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -712,24 +742,43 @@ public void rollbackInflightCompaction(HoodieInstant inflightInstant, HoodieTabl
+   * Rollback all failed commits.
    */
-  private void rollbackPendingCommits() {
+  public void rollbackFailedCommits() {
```
Intention is to make `rollback()` private, since this is the first release, I've only marked it as @deprecated. Going forward people should use the following 2 API's : `restore()` to revert successful commits and `rollbackFailedCommits()` to delete failed writes.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -232,17 +254,18 @@ void emitCommitMetrics(String instantTime, HoodieCommitMetadata metadata, String
-    if (rollbackPending) {
-      rollBackInflightBootstrap();
-    }
+    // TODO : MULTIWRITER -> check if failed bootstrap files can be cleaned later
```
Yes, but now I'm making cleaning/rolling back optional during that reattempt, so want to make sure it's OK to clean/delete later..

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -875,5 +924,6 @@ public void close() {
     // Calling this here releases any resources used by your index, so make sure to finish any related operations
     // before this point
     this.index.close();
+    this.hoodieHeartbeatClient.stop();
```
Did not follow, can you explain what you mean by "differentiate from other usage" ?

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -615,6 +640,11 @@ public String startCommit() {
    * @param instantTime Instant time to be generated
    */
   public void startCommitWithTime(String instantTime) {
+    if (config.getFailedWritesCleanPolicy() == HoodieCleaningPolicy.HoodieFailedWritesCleaningPolicy.LAZY) {
```
Added a private method

n3nash commented:
@vinothchandar Addressed most your comments, replied to the others. At a high level, LAZY rollback is supported only for commits/deltacommits since there is contention when creating multiple writers. Since compactions/clustering don't have contention, they continue to do inline rollback which does not cause any issues -> reason is because they re-use the same instant time so changing them requires deeper changes which is not needed for multi-writer.

n3nash commented:
@vinothchandar ready for review

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -707,24 +739,51 @@ public void rollbackInflightCompaction(HoodieInstant inflightInstant, HoodieTabl
+      // Delete the heartbeats from DFS
+      if (!isEager(config.getFailedWritesCleanPolicy())) {
+        try {
+          this.heartbeatClient.delete(instant);
```
how do we cleanup if we fail after rolling back, but before deleting the heartbeat?

n3nash replied:
Added a Utils class to clean up orphaned heartbeat files

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -859,6 +931,10 @@ protected void setWriteSchemaForDeletes(HoodieTableMetaClient metaClient) {
     }
   }
 
+  private boolean isEager(HoodieFailedWritesCleaningPolicy cleaningPolicy) {
```
Like I mentioned above, we can make turn this into `eagerCleanFailedWrites()` returning boolean, working off the member variable `config` ? In its current form, this should probably reside in `HoodieFailedWritesCleaningPolicy` itself

n3nash replied:
Responded above

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,265 @@
+  // heartbeat interval in millis
+  private final Long heartbeatIntervalInMs;
+  private Integer numTolerableHeartbeatMisses;
+  private final Long maxAllowableHeartbeatInternalInMs;
```
typo: interval

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,265 @@
+    private Boolean isHeartbeatStarted = false;
+    private Boolean isHeartbeatStopped = false;
+    private Long lastHeartBeatTime;
+    private Integer numHeartBeatsForCurrentInstantTime = 0;
```
just `numHeartBeats`?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,265 @@
+   */
+  public void start(String instantTime) {
+    LOG.info("Received request to start heartbeat for instant time " + instantTime);
+    Heartbeat heartbeat = instantToHeartbeatMap.get(instantTime);
```
is there a reason to make this map synchronzied/concurrent ?

n3nash replied:
It's not synchronized, if you see above, it's just a HashMap()

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,265 @@
+    instantToHeartbeatMap.values().stream().forEach(heartbeat -> stop(heartbeat.getInstantTime()));
+  }
+
+  public boolean delete(String instantTime) throws IOException {
```
I find this a bit odd, for heartBeat to have a `delete()` method? can this clean up be done as part of `stop()` ? esp given delete() requires a stop first

n3nash replied:
yeah, I agree, we cannot clean up as part of `stop` since we need the heartbeat later when the `cleanFailedWrites` kicks in and then that method does the clean up of heartbeat. I can move this to a utils class if that makes sense.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieCompactionConfig.java:
```diff
@@ -260,6 +264,11 @@ public Builder withCleanBootstrapBaseFileEnabled(Boolean cleanBootstrapSourceFil
       return this;
     }
 
+    public Builder withCleanFailedWritesPolicy(HoodieFailedWritesCleaningPolicy failedWritesPolicy) {
```
withFailedWritesCleaningPolicy(). can we make this consistent everywhere.

n3nash replied:
Done, is there an easy way to generate a with... for these variables ? I'm manually typing them hence end up introducing errors.

vinothchandar commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/heartbeat/TestHoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,152 @@
+        new HoodieHeartbeatClient(metaClient.getFs(), metaClient.getBasePath(), heartBeatInterval, numTolerableMisses);
+    hoodieHeartbeatClient.start(instantTime1);
+    hoodieHeartbeatClient.stop(instantTime1);
+    await().atMost(2, SECONDS).until(() -> hoodieHeartbeatClient.getHeartbeat(instantTime1).getNumHeartBeatsForCurrentInstantTime() > 0);
```
the 2 second limit is a bit tight. the test jvm can gc etc. can we make this longer.  same check for other time based tests here.

n3nash replied:
I thought await() doesn't necessary wait for 2 SECONDS, it just simulates that internally ?

vinothchandar commented on hudi-client/hudi-spark-client/pom.xml:
```diff
@@ -127,6 +127,14 @@
       <scope>test</scope>
     </dependency>
 
+    <!-- Other Utils -->
```
can we pull this into parent pom, introduce a property for the version and just use this without the `<version>` tag here. just like other deps/

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieFailedWritesCleaningPolicy.java:
```diff
@@ -0,0 +1,23 @@
+package org.apache.hudi.common.model;
+
+public enum HoodieFailedWritesCleaningPolicy {
+  EAGER, LAZY, NEVER
```
docs for all these three modes please.

vinothchandar reviewed (commented):
Overall looks good to me. 
Two major things. 

1) We need to introduce a new config `hoodie.table.services.enable` which should be `true` by default. and we guard all of the following below during `postCommit` or `postWrite` using that flag. i.e only do it if the flag is true. 
 - scheduling of cleaning, compaction, clustering
 - inline cleaning, compaction, clustering
 - archival
the expectation is that users will have to turn the flag off for the other writers. 

2. Can we make a pass at all points where compaction (and clustering) is triggered (cli, the HoodieCompactor tool, async compaction runners in data source,deltastreamer) and ensure the policy is set to `EAGER` .

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -232,17 +250,18 @@ void emitCommitMetrics(String instantTime, HoodieCommitMetadata metadata, String
-    }
+    // TODO : MULTIWRITER -> check if failed bootstrap files can be cleaned later
     HoodieTable<T, I, K, O> table = getTableAndInitCtx(WriteOperationType.UPSERT, HoodieTimeline.METADATA_BOOTSTRAP_INSTANT_TS);
+    if (isEager(config.getFailedWritesCleanPolicy())) {
```
The `isEager` is needed for multiple operations, not just `rollbackFailedBootstrap()`, changed it to do check within method but cannot do operations inside eagerCleanFailedWrites. This will need more refactoring (if possible), I can open a follow up ticket and take it up right after the release to adhere against release timeline.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -580,6 +602,11 @@ public HoodieRestoreMetadata restoreToInstant(final String instantTime) throws H
     LOG.info("Cleaner started");
     final Timer.Context timerContext = metrics.getCleanCtx();
+    LOG.info("Cleaning failed attempts if any");
+    if (!isEager(config.getFailedWritesCleanPolicy())) {
```
Changed the method yes, but cannot pull the LOG inside for reasons stated above...

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -408,13 +426,14 @@ protected void postCommit(HoodieTable<T, I, K, O> table, HoodieCommitMetadata me
 
   protected void runAnyPendingCompactions(HoodieTable<T, I, K, O> table) {
-    table.getActiveTimeline().getCommitsAndCompactionTimeline().filterPendingCompactionTimeline().getInstants()
+    table.getActiveTimeline().getCommitsCompactionAndReplaceTimeline().filterPendingCompactionTimeline().getInstants()
```
yes, I think I renamed it in the other PR, since this is landing first, renamed it here

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -707,24 +739,51 @@ public void rollbackInflightCompaction(HoodieInstant inflightInstant, HoodieTabl
-        .collect(Collectors.toList());
-    for (String commit : commits) {
-      if (HoodieTimeline.compareTimestamps(commit, HoodieTimeline.LESSER_THAN_OR_EQUALS,
+    List<String> instantsToRollback = getInstantsToRollback(table);
```
For part 1 of the question, what you mentioned is true, in addition, I have also added a check in the merging of log blocks where if the instant is not on the active timeline, that block will be ignored (so for this specific case, rollback blocks are not needed). For the case where the instant has been archived, that cannot happen until that failed commit has been rolled back by the cleaner (changes to the archival ensure that)

For part 2, it's slightly complicated. 

For parallel writing, the assumption is the following : 
1) Writer 1 starts writing to file groups f1_c1, f2_c1
2) Writer 2 starts writing to file groups f3_c2, f4_c2
3) Any scheduling operations (schedule compaction, schedule clustering, clean) cannot happen for f1, f2, f3, f4 because updates may have started for f1.log_c1 and scheduleCompaction may create a new phantom file slice f1_c3, in which case f1.log_c1 updates are lost. 
So, for all operations to run concurrently, they truly have to not overlap in file slices

For concurrent writing, we may need some more changes :
1) Writer 1 starts writing to file groups f1_c1, f2_c1
2) Writer 2 starts writing to file groups **f2_c2**, f3_c2
3) scheduleCompaction for f2_c1 & f1_c1 -> f2_c3, f1_c3

Now the other PR needs some more enhancements to not only check how many writers have happened since the time it started, but also if there are any other "schedule" operations that are also on the timeline. This will I think we will have to implement a ConcurrentRequestStrategyWithPriority where we can allow for schedule to fail if it sees a writer running etc..

n3nash replied:
As per discussion offline, we are just considering this PR to disable inline rollbacks, parallel writing will NOT be supported

vinothchandar replied:
can you comment again on whether this changes the answer to my question above? 

> I have also added a check in the merging of log blocks where if the instant is not on the active timeline, that block will be ignored 

The concern I had was the part 2 where, a committed write could have been archived and we may end up skipping it. Can you please clarify again how we guard that? By ensuring the archival will wait for the cleaner to log this block?

n3nash commented:
@vinothchandar This PR is ready, concurrency control added for writers and table services.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,265 @@
+import java.util.stream.Collectors;
+
+/**
+ * This class creates heartbeat for hudi client. This heartbeat is used to ascertain whether the running job is or not.
```
@n3nash spent sometime thinking about this. the tradeoffs are pretty clear in the sense, this helps us reclaim the storage used by failed writes much sooner than say deleting the files, during archival process (which by itself is not a guarantee actually).

If we want to implement some schema to pre-emptively take away compute resources used by a pending job or storage space, then I don't see a way to do this without heartbeats of some kind. 

We could just keep this simpler in this PR, if you like and just clean up once during archival. (again, the case of a long deletion job running for say 12 hrs, comes to mind. and That needs some kind of heartbeat mechanism, so that even archival does not clean up). 

Orthogonal question : for such long running concurrent writes, should n't they just abort early if they detect more commits have happened since it began?  i.e if it figures out after 1 hr of runtime, that some writes have changed the same filegroups, then why should it even run? its going to fail anyway

vinothchandar replied:
will review this class again, pending this discussion here.

n3nash replied:
High level, no other way to guard against these cases. Even if we check for long running concurrent writes and resolve every N heartbeats, backfill jobs may not have anything overlapping and may even OOM at the end since folks don't know how to tune large spark jobs.

vinothchandar commented:
@n3nash rebase, repush and get CI to be happy?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -160,7 +180,8 @@ public boolean commit(String instantTime, O writeStatuses) {
     HoodieTableMetaClient metaClient = createMetaClient(false);
     String actionType = metaClient.getCommitActionType();
-    return commit(instantTime, writeStatuses, extraMetadata, actionType, Collections.emptyMap());
+    boolean isCommitted = commit(instantTime, writeStatuses, extraMetadata, actionType, Collections.emptyMap());
```
remote `isCommitted` and return directly?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -732,6 +746,9 @@ public void rollbackInflightCompaction(HoodieInstant inflightInstant, HoodieTabl
   }
 
   /**
+<<<<<<< 1e1a9a5ba5907efb5dd8cb1e0ae830aa78725dd1
```
this seems like remnant of merge conficts? can you please make a full pass and ensure no temp debugging changes/ or remnants like this are removed?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -750,24 +767,49 @@ private HoodieTimeline getInflightTimelineExcludeCompactionAndClustering(HoodieT
   /**
-   * Cleanup all pending commits.
+   * Clean up all failed commits.
+=======
```
same here

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -750,24 +767,49 @@ private HoodieTimeline getInflightTimelineExcludeCompactionAndClustering(HoodieT
+      HoodieTimeline inflightTimeline = table.getMetaClient().getCommitsTimeline().filterPendingExcludingCompaction();
+      return inflightTimeline.getReverseOrderedInstants().map(HoodieInstant::getTimestamp)
+          .collect(Collectors.toList());
+    } else if (!config.getFailedWritesCleanPolicy().isEager()) {
```
would nt `HoodieFailedWritesCleaningPolicy.NEVER` also pass the check for `!eager` ?

n3nash replied:
moved the else if conditions

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -859,6 +901,19 @@ protected void finalizeWrite(HoodieTable<T, I, K, O> table, String instantTime,
     }
   }
 
+  protected void checkHeartbeatExpired(String instantTime, HoodieTable table) {
```
lets move this to `HeartbeatUtils`

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -133,6 +134,12 @@
+  public static final String CLIENT_HEARTBEAT_INTERVAL_IN_MS_PROP = "hoodie.client.heartbeat.interval_in_ms";
+  public static final Integer DEFAULT_CLIENT_HEARTBEAT_INTERVAL_IN_MS = 60 * 1000;
+
+  public static final String CLIENT_HEARTBEAT_NUM_TOLERABLE_MISSES_PROP = "hoodie.client.heartbeat.tolerable_misses";
```
nit: tolerable.misses

vinothchandar commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/heartbeat/TestHoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,145 @@
+    th2.start();
+    th3.start();
+
+    // To simulate the issues of race conditions when creating & reading file, simply do th1.join; th2.join; th3.join;
```
was wondering how these threads actually stop?

n3nash replied:
I left the test disabled since I was just using this for testing. Have enabled the test now and added the join back.

vinothchandar commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/heartbeat/TestHoodieHeartbeatClient.java:
```diff
@@ -0,0 +1,145 @@
+        }
+      } catch (Exception e) {
+        LOG.info("Caught Exception th3");
+        e.printStackTrace();
```
`LOG.error()` and get rid of printing stack trace?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -105,7 +112,8 @@ public static SparkConf registerClasses(SparkConf conf) {
                         String commitActionType, Map<String, List<String>> partitionToReplacedFileIds) {
     List<HoodieWriteStat> writeStats = writeStatuses.map(WriteStatus::getStat).collect();
-    return commitStats(instantTime, writeStats, extraMetadata, commitActionType, partitionToReplacedFileIds);
+    boolean isCommitted = commitStats(instantTime, writeStats, extraMetadata, commitActionType, partitionToReplacedFileIds);
```
same. lets remove the temp variable

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/HeartbeatUtils.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Helper class to delete heartbeat for completed or failed instants with expired heartbeats.
+ */
+public class HeartbeatUtils {
```
is there a reason why this should be in `hudi-common`?

n3nash replied:
No, I thought I created it in `hudi-client-common`, moved it.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/HeartbeatUtils.java:
```diff
@@ -0,0 +1,78 @@
+   * @param metaClient
+   * @param basePath
+   */
+  public static void cleanUpExpiredOrOrphanHeartbeatFiles(List<String> existingHeartbeatInstants,
```
rename by dropping  the `Files` suffix? is there any special handling for expired vs orphaned? why not just call it `cleanExpiredHearbeats`

n3nash replied:
There isn't any special handling, renamed it.

vinothchandar reviewed (commented):
Few comments. High level, LGTM. 

can we also add a test with metadatatable enabled and actually do some production runs?

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -128,11 +133,26 @@ public AbstractHoodieWriteClient(HoodieEngineContext context, HoodieWriteConfig
    * @param timelineService Timeline Service that runs as part of write client.
    */
+  @Deprecated
   public AbstractHoodieWriteClient(HoodieEngineContext context, HoodieWriteConfig writeConfig, boolean rollbackPending,
```
That would be a backwards incompatible change. I'm OK to remove it if it's just for Uber.

vinothchandar replied:
yeah nobody uses the RDD api other than Uber. so please remove.

n3nash commented:
> The concern I had was the part 2 where, a committed write could have been archived and we may end up skipping it. Can you please clarify again how we guard that? By ensuring the archival will wait for the cleaner to log this block?

@vinothchandar I have made 2 changes for multi-writer. 

1. I changed the logic in `HoodieTimelineArchiveLog` to NOT archive anything after the oldest inflight. For long running jobs, we want to make sure that all instants that happened after it are present and not archived so conflict resolution can be done correctly. 
2. I have added a check in log scanning code that log blocks instant should be either a) in the commit timeline or earlier that the earliest commit and b) not present in the inflight timeline. 

Not following your ask around committed writes and skipping, can you elaborate ?

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -236,17 +226,16 @@ protected void syncTableMetadata() {
+   * Main API to rollback failed bootstrap.
    */
-  protected void rollBackInflightBootstrap() {
+  public void rollbackFailedBootstrap() {
```
We are deprecating the use pf `rollback` API since that api currently supports rolling back failed and valid commits. Rollback of valid commits is only supported through `restore`.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -424,15 +411,15 @@ protected void postCommit(HoodieTable<T, I, K, O> table, HoodieCommitMetadata me
       autoCleanOnCommit();
-
       syncTableMetadata();
+      this.heartbeatClient.stop(instantTime);
```
Good point, moved it.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -617,11 +610,8 @@ public HoodieCleanMetadata clean() {
-      rollbackPendingCommits();
-    }
+    CleanerUtils.rollbackFailedWrites(config.getFailedWritesCleanPolicy(),
+        HoodieTimeline.COMMIT_ACTION, () -> rollbackFailedWrites());
```
Let's take this offline. We can probably refactor but not in this PR.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HeartbeatUtils.java:
```diff
@@ -0,0 +1,95 @@
+   */
+  public static void cleanExpiredHeartbeats(List<String> existingHeartbeatInstants,
+                                            HoodieTableMetaClient metaClient, String basePath) {
+    Set<String> heartbeatInstants = metaClient.getActiveTimeline()
```
renamed to `nonExpiredHeartbeatInstants` since "active" can mean differnet

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -749,22 +739,49 @@ private HoodieTimeline getInflightTimelineExcludeCompactionAndClustering(HoodieT
+   * Rollback all failed writes.
    */
-  private void rollbackPendingCommits() {
+  public Boolean rollbackFailedWrites() {
```
This function is used as a Lambda in a different place and implemented as a Function0<Boolean>. Hence returning boolean here.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -912,5 +929,6 @@ public void close() {
     // Calling this here releases any resources used by your index, so make sure to finish any related operations
     // before this point
     this.index.close();
+    this.heartbeatClient.stop();
```
This is to stop all heartbeats if multiple were started when the client goes out of scope.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HeartbeatUtils.java:
```diff
@@ -0,0 +1,95 @@
+        LOG.error("Failed to delete heartbeat for instant " + instantTime);
+      }
+    } catch (IOException io) {
+      LOG.error("Unable to delete heartbeat for instant " + instantTime, io);
```
No, not being able to delete heartbeat is benign, should not be a drastic failure for job.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HeartbeatUtils.java:
```diff
@@ -0,0 +1,95 @@
+      String heartbeatFolderPath = HoodieTableMetaClient.getHeartbeatFolderPath(basePath);
+      deleted = fs.delete(new Path(heartbeatFolderPath + File.separator + instantTime), false);
+      if (!deleted) {
+        LOG.error("Failed to delete heartbeat for instant " + instantTime);
```
No, sometimes the underlying DFS may not delete the file due to an exception. Unfortunately, the `delete` API does not return an exception. We are logging the error here as opposed to throwing an exception since not being able to delete a heartbeat is benign and repeatable.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -598,6 +588,9 @@ public HoodieRestoreMetadata restoreToInstant(final String instantTime) throws H
     LOG.info("Cleaner started");
     final Timer.Context timerContext = metrics.getCleanCtx();
+    LOG.info("Cleaned failed attempts if any");
+    CleanerUtils.rollbackFailedWrites(config.getFailedWritesCleanPolicy(),
```
Yes, it is handled in `BaseRollbackActionExecution`

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -236,17 +226,16 @@ protected void syncTableMetadata() {
-    }
+    // TODO : MULTIWRITER -> check if failed bootstrap files can be cleaned later
     HoodieTable<T, I, K, O> table = getTableAndInitCtx(WriteOperationType.UPSERT, HoodieTimeline.METADATA_BOOTSTRAP_INSTANT_TS);
+    rollbackFailedBootstrap();
```
This API is only supported via JavaRDD which is used only at Uber. No side-effect for others.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -181,7 +171,7 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats, Opti
     // Finalize write
     finalizeWrite(table, instantTime, stats);
-
+    HeartbeatUtils.abortIfHeartbeatExpired(instantTime, table, heartbeatClient, config);
```
I want to avoid business logic like approach 2 in the `AbstractHoodieWriteClient`, this logic is pushed to the method `abortIfHeartbeatExpired`. Alternatively, the method `abortIfHeartbeatExpired` build on other public API's to be able to do custom business logic if needed.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -749,22 +739,49 @@ private HoodieTimeline getInflightTimelineExcludeCompactionAndClustering(HoodieT
+  }
+
+  private List<String> getInstantsToRollback(HoodieTable<T, I, K, O> table) {
+    if (config.getFailedWritesCleanPolicy().isEager()) {
```
Would like to keep if-else since I don't see the types of policies growing at this point.

n3nash commented:
@nsivabalan Addressed your comments PTAL. @vinothchandar This PR is ready. Following is the summary of outstanding items / previous questions. 
I have tested this PR against the test plan discussed. I have not been able to test against async clustering / compaction. This should be technically fine since there are no code changes to the rollback of their path. Let me know if you need any further clarification, we can do a quick call to resolve and land this PR.

> > The concern I had was the part 2 where, a committed write could have been archived and we may end up skipping it. Can you please clarify again how we guard that? By ensuring the archival will wait for the cleaner to log this block?
> 
> @vinothchandar I have made 2 changes for multi-writer.
> 
> 1. I changed the logic in `HoodieTimelineArchiveLog` to NOT archive anything after the oldest inflight. For long running jobs, we want to make sure that all instants that happened after it are present and not archived so conflict resolution can be done correctly.
> 2. I have added a check in log scanning code that log blocks instant should be either a) in the commit timeline or earlier that the earliest commit and b) not present in the inflight timeline.
> 
> Not following your ask around committed writes and skipping, can you elaborate ? Resolved other comments, left 1 open item.

nsivabalan commented:
IIUC, we are delegating the clean up of failed writes to cleaner. Did you folks discuss the possibility of introducing another async table service called FailedJobsCleaner or something. Bcoz, we can't force customer to make regular cleaning aggressive as they might have their own reasons to retain for longer period of time. But what if they wish to clean up failed writes sooner but not inline ;) Also, from my understanding, this new "FailedJobsCleaner" does not need to coordinate w/ anything or take locks. just need to delete unreachable files. So, overhead may not be much as compared to other jobs. no updates to metadata as well right? or do we need to? 

I understand we keep adding more weight to async operations (cleaning, compaction, clustering, etc). anyways, wanted to hear your thoughts.

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/io/TestHoodieTimelineArchiveLog.java:
```diff
@@ -408,11 +408,11 @@ public void testArchiveCommitCompactionNoHole() throws IOException {
     HoodieTimelineArchiveLog archiveLog = new HoodieTimelineArchiveLog(cfg, table);
 
-    HoodieTimeline timeline = metaClient.getActiveTimeline().getCommitsAndCompactionTimeline();
+    HoodieTimeline timeline = metaClient.getActiveTimeline().getWriteTimeline();
```
can we add a test case for archiving with inflight instants and lazy cleaning?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -71,10 +72,10 @@
-  public static final String AUXILIARYFOLDER_NAME = METAFOLDER_NAME + Path.SEPARATOR + ".aux";
-  public static final String BOOTSTRAP_INDEX_ROOT_FOLDER_PATH = AUXILIARYFOLDER_NAME + Path.SEPARATOR + ".bootstrap";
-
+  public static final String TEMPFOLDER_NAME = METAFOLDER_NAME + File.separator + ".temp";
```
why this change?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordScanner.java:
```diff
@@ -145,6 +148,14 @@ public void scan() {
           // hit a block with instant time greater than should be processed, stop processing further
           break;
         }
+        if (r.getBlockType() != CORRUPT_BLOCK && r.getBlockType() != COMMAND_BLOCK) {
```
test case for this?

vinothchandar reviewed (approved):
LGTM. @n3nash I am going to land this after CI passed. I already rebased this. 

Left some minor comments and suggestions for tests. We can add this to the subsequent PR.
